### PR TITLE
Email subsampling and changes in preprocessing logic

### PR DESCRIPTION
**Pull request by mohit-surana**
1. `dal.get_emails` now has an additional parameter called `max_users` which lets us run the model on a subset of data filtered by user instead of arbitrarily picked the first N emails
2. A new function `__clean_data_glove` has been added to tokenize the words based on the format followed by GloVe embeddings. This requires a dependency that has been pushed as a resource.
3. It was noticed that additional mails in an email thread, contrary to our understanding, had not been removed which led to extremely long email bodies.

Changes 1 and 3 should improve the training process considerably as there will be less number of mails (faster training) which are more relevant (better performance) as well.

Future enhancement: Change 3 could be refactored to work with w2v custom and the previous `__clean_data` implementation as well

### CODE REVIEW

**Comment by mohit-surana** [comment]
@mudit1990 
After this is reviewed and merged, you will need to modify line 72 in dal.py to use the new cleaning logic.
Can you update the csv files once done?

File: code/dal.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
+import random
 import string
 import time
+import utils
 
+from nltk.tokenize.stanford import StanfordTokenizer
 
-def get_emails(num_emails=100, fetch_all=False):
+
+def get_emails(num_emails=100, max_users=150, fetch_all=False):
```
Comment (mudit1990) [inline code/dal.py:12]: I don't think its worth having num_emails and max_users together. We should keep only max_users
If someone asks num_emails to be 5k, then this code will select the smallest file (e.g 5k_emails.csv or 10k_emails.csv) that satisfies the requirement effectively negating the work of max_users

Or even better, we can have a separate get_emails(max_users) that always reads from the entire dataset and filters based on the number of users required

Comment (mudit1990) [reply code/dal.py:12]: Plus it will be much simpler and more intuitive if we return the caller exactly the number of users requested

File: code/dal.py [hunk 2]
```diff
@@ -94,6 +121,34 @@ def __clean_data(data):
     return data
 
 
+def __clean_data_glove(data):
```
Comment (mudit1990) [inline code/dal.py:124]: how is this related to glove vocabulary? does it mean that this method cannot be used when we train our custom word2vec?

Comment (mohit-surana) [reply code/dal.py:124]: The words in GloVe are tokenized using the StanfordParser which behaves a bit differently compared to naive replacing of punctuations and splitting
Eg. can't is tokenized as ca + n't and punctuations have their word embeddings as well

We can definitely use the same for custom word2vec as well
But are not constrained by it

Perhaps the function can be renamed to just __clean_data once we deprecate the old version.

File: code/dal.py (hunk 1 above)
Comment (mohit-surana) [reply code/dal.py:12]: Yes, these are conflicting requirements.

Going forward, should we retain the num_emails option (two functions)? 
Or is it enough to just have max_users?

**Review by mudit1990 — APPROVED** [review]

**Comment by mudit1990** [comment]
merging this for now ... as per our discussion, the get_emails will be broken up into 2 functions, get_emails(returns fixed number of emails) and get_emails_byusers(returns data for fixed number of users)
